pkg/revisions/tag_watcher.go

```diff
@@ -0,0 +1,126 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package revisions
+
+import (
+	"sync"
+
+	admissionregistrationv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"istio.io/api/label"
+	"istio.io/istio/istioctl/pkg/tag"
+	"istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/kube/controllers"
+	"istio.io/istio/pkg/kube/kclient"
+	"istio.io/istio/pkg/kube/kubetypes"
+	"istio.io/pkg/log"
+)
+
+// TagWatcher keeps track of the current tags and can notify watchers
+// when the tags change.
+type TagWatcher interface {
+	Run(stopCh <-chan struct{})
+	HasSynced() bool
+	AddHandler(handler TagHandler)
+}
+
+// TagHandler is a callback for when the tags revision change.
+type TagHandler func([]string)
+
+type tagWatcher struct {
+	revision string
+	handlers []TagHandler
+
+	queue           controllers.Queue
+	webhookInformer kclient.Client[*admissionregistrationv1.MutatingWebhookConfiguration]
+	mu              sync.RWMutex
+	tagsToRevisions map[string]string
+	revisionsToTags map[string][]string
+}
+
+func NewTagWatcher(client kube.Client, revision string) TagWatcher {
+	p := &tagWatcher{
+		revision:        revision,
+		mu:              sync.RWMutex{},
+		tagsToRevisions: map[string]string{},
+		revisionsToTags: map[string][]string{},
+	}
+	p.queue = controllers.NewQueue("tag", controllers.WithReconciler(p.updateTags))
+	p.webhookInformer = kclient.NewFiltered[*admissionregistrationv1.MutatingWebhookConfiguration](client, kubetypes.Filter{
+		ObjectFilter:    isTagWebhook,
+		ObjectTransform: kube.StripUnusedFields,
```

Comment:
nit: this is default, not needed

Reply:
Can you clarify what is default?  The UI is showing four lines of code, and I am fairly confident some of them are not default...

Reply:
kube.StripUnusedFields